src/nunavut/lang/cpp/support/VariableLengthArray.hpp

```diff
@@ -0,0 +1,464 @@
+    {
+        return alloc_;
+    }
+
+    // +----------------------------------------------------------------------+
+    // | CAPACITY
+    // +----------------------------------------------------------------------+
+    /**
+     * The number of elements that can be stored in the array without further
+     * allocations. This number will only grow through calls to {@code reserve}
+     * and can shrink through calls to {@code shrink_to_fit}. This value shall
+     * never exceed {@code max_size}.
+     */
+    constexpr std::size_t capacity() const noexcept
+    {
+        return capacity_;
+    }
+
+    /**
+     * The current number of elements in the array. This number increases with each
+     * successful call to {@code push_back_no_alloc} and decreases with each call to
+     * {@code pop_back} (when size is > 0).
+     */
+    constexpr std::size_t size() const noexcept
+    {
+        return size_;
+    }
+
+    /**
+     * Ensure enough memory is allocated to store at least the {@code desired_capacity} number of elements.
+     * This container is different then STL vector in requiring up-front reservation of the necessary
+     * capacity. It does not allocate on push_back.
+     *
+     * @param  desired_capacity The number of elements to allocate, but not initialize, memory for.
+     * @return The new (or unchanged) capacity of this object.
+     */
+    std::size_t reserve(const std::size_t desired_capacity) noexcept(
+        noexcept(
+            allocator_type().allocate(0)
+        )
+        &&
+        noexcept(
+            move_and_free<T>(nullptr, nullptr, 0, 0, std::declval<Allocator&>())
+        )
+    )
+    {
+        const std::size_t clamped_capacity   = (desired_capacity > MaxSize) ? MaxSize : desired_capacity;
+        const std::size_t no_shrink_capacity = (clamped_capacity > size_) ? clamped_capacity : size_;
+
+        T* new_data = nullptr;
+        try
+        {
+            new_data = alloc_.allocate(no_shrink_capacity);
+        }
+        catch (const std::bad_alloc& e)
+        {
+            // we ignore the exception since all allocation failures are modeled using
+            // null by this class.
+            (void) e;
+        }
+
+        if (new_data != nullptr)
+        {
+            if (new_data != data_)
+            {
+                move_and_free(new_data, data_, size_, capacity_, alloc_);
+            }  // else the allocator was able to simply extend the reserved area for the same memory pointer.
+            data_     = new_data;
+            capacity_ = no_shrink_capacity;
+        }
+
+        return capacity_;
+    }
+
+    /**
+     * Deallocate or reallocate memory such that not more than {@code size()} elements can be stored in this object.
+     *
+     * @return True if nothing was done or if memory was successfully resized. False to indicate that the allocator
+     *         could not provide enough memory to move the existing objects to a smaller allocation.
```

Comment:
> False to indicate that the allocator could not provide enough memory to move the existing objects to a smaller allocation

Is there any practical scenario where this information is actionable? How do you envision this method to be used? My understanding is that it may be leveraged in some very specific cases involving non-hard-real-time processes that desire to conserve memory. If the memory could not be reclaimed, from the standpoint of such process it is not an error since it does not affect the business logic (the data is still stored and reachable). If my understanding is correct, then probably every invocation of this method will have to be accompanied with an explicit cast to void, so maybe it should return nothing?

Reply:
An error occurred so it has to be communicated out. If you choose to void cast it away you ignore the error but that is a choice the application has to make not the framework.